src/Signature.php

```diff
@@ -30,9 +30,21 @@ public function __construct($message, Key $key, AlgorithmInterface $algorithm, H
 
     public function string()
     {
-        return $this->algorithm->sign(
-            $this->key->secret,
-            $this->signingString->string()
-        );
+        switch (get_class($this->algorithm)) {
+          case 'HttpSignatures\HmacAlgorithm':
+            return $this->algorithm->sign(
+                $this->key->secret,
+                $this->signingString->string()
+            );
+          case 'HttpSignatures\RsaAlgorithm':
+            return $this->algorithm->sign(
+                $this->key->privateKey,
+                $this->signingString->string()
+              );
+          default:
+            throw new Exception(
+              "Algorithm class " .
+               get_class($this->algorithm) . " unknown");
+        }
```

Comment:
ideally this code wouldn't exist - use OO, not switch

Reply:
Maybe a generic:
```php
public function getSigningKey()
public function getVirifierKey()
```

For PKI (RSA/EC), different responses, for HMAC the same object (a shared secret)?

Reply:
Yup, above approach much simpler removing switches and requirements for key consumers to know about key details. Key state now private, keys are returned by purpose instead of explicit attributes (should make it more trivial to extend to EC too). 2bacdc7e361deeb92c97ab103a1479f8a7274445